grpc/shared/container.go

```diff
@@ -0,0 +1,207 @@
+
+	// auctioneer address is a bech32m address
+	auctioneerAddress atomic.Pointer[string]
+	// this is set to the height at which the first auctioneer address is activated.
+	// before `auctioneerStartHeight` any incoming `Allocations` will be ignored
+	auctioneerStartHeight uint64
+
+	nextFeeRecipient atomic.Pointer[common.Address] // Fee recipient for the next block
+}
+
+func NewSharedServiceContainer(eth *eth.Ethereum) (*SharedServiceContainer, error) {
+	bc := eth.BlockChain()
+
+	if bc.Config().AstriaRollupName == "" {
+		return nil, errors.New("rollup name not set")
+	}
+
+	if bc.Config().AstriaSequencerInitialHeight == 0 {
+		return nil, errors.New("sequencer initial height not set")
+	}
+
+	if bc.Config().AstriaCelestiaInitialHeight == 0 {
+		return nil, errors.New("celestia initial height not set")
+	}
+
+	if bc.Config().AstriaCelestiaHeightVariance == 0 {
+		return nil, errors.New("celestia height variance not set")
+	}
+
+	bridgeAddresses := make(map[string]*params.AstriaBridgeAddressConfig)
+	bridgeAllowedAssets := make(map[string]struct{})
+	if bc.Config().AstriaBridgeAddressConfigs == nil {
+		log.Warn("bridge addresses not set")
+	} else {
+		nativeBridgeSeen := false
+		for _, cfg := range bc.Config().AstriaBridgeAddressConfigs {
+			err := cfg.Validate(bc.Config().AstriaSequencerAddressPrefix)
+			if err != nil {
+				return nil, fmt.Errorf("invalid bridge address config: %w", err)
+			}
+
+			if cfg.Erc20Asset == nil {
+				if nativeBridgeSeen {
+					return nil, errors.New("only one native bridge address is allowed")
+				}
+				nativeBridgeSeen = true
+			}
+
+			if cfg.Erc20Asset != nil && cfg.SenderAddress == (common.Address{}) {
+				return nil, errors.New("astria bridge sender address must be set for bridged ERC20 assets")
+			}
+
+			bridgeCfg := cfg
+			bridgeAddresses[cfg.BridgeAddress] = &bridgeCfg
+			bridgeAllowedAssets[cfg.AssetDenom] = struct{}{}
+			if cfg.Erc20Asset == nil {
+				log.Info("bridge for sequencer native asset initialized", "bridgeAddress", cfg.BridgeAddress, "assetDenom", cfg.AssetDenom)
+			} else {
+				log.Info("bridge for ERC20 asset initialized", "bridgeAddress", cfg.BridgeAddress, "assetDenom", cfg.AssetDenom, "contractAddress", cfg.Erc20Asset.ContractAddress)
+			}
+		}
+	}
+
+	// To decrease compute cost, we identify the next fee recipient at the start
+	// and update it as we execute blocks.
+	nextFeeRecipient := common.Address{}
+	nextBlock := uint32(bc.CurrentBlock().Number.Int64()) + 1
+	if bc.Config().AstriaFeeCollectors == nil {
+		log.Warn("fee asset collectors not set, assets will be burned")
+	} else {
+		maxHeightCollectorMatch := uint32(0)
+		for height, collector := range bc.Config().AstriaFeeCollectors {
+			if height <= nextBlock && height > maxHeightCollectorMatch {
+				maxHeightCollectorMatch = height
+				nextFeeRecipient = collector
+			}
+		}
+	}
+
+	if bc.Config().AstriaAuctioneerAddresses == nil {
+		return nil, errors.New("auctioneer addresses not set")
+	}
```

Comment:
nit: I think ideally it's ok if we haven't set an auctioneer address, allows us to test without and then add in on the binary, similar to the above block on fee collectors, just log out a warn or an info.

Noting this logic probably changes quite a bit after the geth updates for genesis.

Reply:
addressed here: https://github.com/astriaorg/flame/pull/30/commits/c969f99dd4cd8788573cb927873c766a2d12533a